File: services/api/app/api/v1/scenario.py
"""GET /api/v1/properties/{bbl}/scenario - internal deterministic scenario
(task M5-T003).

SECURITY / DEPLOYMENT STATUS: INTERNAL/DEV ONLY, and additionally FEATURE-FLAG
GATED OFF BY DEFAULT - exactly like the rule-evaluation route this mirrors. Two
independent guards keep it off in production:

1. Like every route in this service it has NO authentication yet (M0-T007/T008
   blocked on the Supabase token); the service must not be publicly exposed.
2. It is reachable ONLY when ``INTERNAL_SCENARIO_ENABLED`` is an explicit true
   token (:func:`app.config.internal_scenario_enabled`). Absent / empty /
   unknown -> DISABLED (fail safe): the handler returns a generic ``404 Not
   Found`` byte-indistinguishable from an unmounted path, leaking NO hint the
   feature exists. The route is registered ``include_in_schema=False`` so it
   never appears in the OpenAPI document regardless of the flag.

What it does (deterministic route; NO independent legal calculation lives here):
rebuild the canonical property profile AND its rule_evaluation SERVER-SIDE over
the SAME trusted, injected seams the accepted rule-evaluation route uses
(``get_pluto_fetcher`` -> ``build_property_profile`` -> ``evaluate_property`` ->
``serialize_rule_evaluation``, with the server-side spatial substrate from
``get_spatial_substrate_provider``), then run the already-built deterministic
``build_scenario`` and return a versioned ``scenario`` @ 1.0.0 document. The
endpoint NEVER accepts a request body or a browser-supplied profile - only the
``bbl`` path parameter - so an untrusted caller can never inject the facts a
scenario would rest on. The only material number a scenario can surface is the
canonical ``max_residential_floor_area_sq_ft`` cap ALREADY present in the
rule_evaluation trace, surfaced VERBATIM by ``build_scenario``; nothing is
recomputed here.

A legitimate no-scenario / unsupported / professional-review outcome is a NORMAL
200 ``scenario`` document (``scenario_kind`` no_scenario / unsupported;
``coverage_status`` professional_review_required / data_conflict / unsupported /
not_applicable), never an error - so a consumer keeps the property usable. Only
genuine faults (malformed BBL, upstream fetch failure, internal contract defect,
unexpected exception) become typed API errors, mapped to the SAME HTTP semantics
as the property and rule-evaluation routes and carrying no traceback, path,
secret, or internal string.

The exact set of emitted (HTTP status, state) pairs is the single source of
truth ``STATUS_STATE_MATRIX`` below; it equals the rule-evaluation route's
documented matrix (the scenario contract defect reuses the shared
``internal_contract_error`` 500 pair, so it adds no new pair).
"""

from __future__ import annotations

import logging
import uuid

from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse

from app.api.v1.properties import (
    _DEFAULT_ERROR_STATUS,
    _ERROR_STATUS,
    PlutoFetcher,
    get_pluto_fetcher,
)
from app.api.v1.rule_evaluation import (
    SpatialSubstrateProvider,
    get_spatial_substrate_provider,
)
from app.config import internal_scenario_enabled
from app.connectors.bbl import BBLValidationError, normalize_bbl
from app.connectors.pluto_soda import DATASET_ID, SOURCE_ID, PlutoConnectorError
from app.profile.builder import build_property_profile
from app.profile.contract import (
    ContractValidationError,
    UnsupportedContractVersionError,
    validate_profile,
)
from app.rules.integration import evaluate_property
from app.rules.response import (
    RuleEvaluationContractError,
    serialize_rule_evaluation,
    validate_rule_evaluation_document,
)
from app.scenario import build_scenario
from app.scenario.contract import ScenarioContractError, validate_scenario_document

__all__ = ["STATUS_STATE_MATRIX", "router"]

logger = logging.getLogger("app.api.v1.scenario")

router = APIRouter(prefix="/api/v1", tags=["scenario"])

# ---------------------------------------------------------------------------
# EXACT (HTTP status, state) pair matrix. Every emission path below is
# enumerated here; it equals the rule-evaluation route's documented matrix
# (a fail-safe no-scenario result is a NORMAL 200 with no ``state`` field - the
# ``None`` sentinel records the exhaustive 200 path). The scenario-document
# contract defect maps to the shared ``internal_contract_error`` 500 pair, so it
# never introduces a status/state pair the mirrored route does not also emit.
# ---------------------------------------------------------------------------
STATUS_STATE_MATRIX: frozenset[tuple[int, str | None]] = frozenset(
    {
        (200, None),  # scenario document (validated before send)
        (422, "validation_error"),  # malformed BBL, no connector call
        (404, "no_match"),  # valid BBL, no PLUTO record (a result)
        (502, "schema_drift"),  # dataset contract breakage
        (503, "rate_limited"),  # SODA throttling after retry budget
        (503, "source_unavailable"),  # SODA outage after retry budget
        (504, "timeout"),  # SODA timeout after retry budget
        (500, "internal_error"),  # unexpected internal defect (generic)
        (500, "internal_contract_error"),  # a built payload failed its contract
    }
)


def _json(status_code: int, body: dict, correlation_id: str) -> JSONResponse:
    return JSONResponse(
        status_code=status_code,
        content=body,
        headers={"X-Correlation-ID": correlation_id},
    )


def _not_found() -> JSONResponse:
    """Generic 404 identical to FastAPI's default for an unmounted path. Carries
    NO correlation id and NO body hint, so a disabled feature is indistinguishable
    from a route that does not exist (fail-safe production disable)."""
    return JSONResponse(status_code=404, content={"detail": "Not Found"})


def _internal_error_500(correlation_id: str) -> JSONResponse:
    """Documented generic 500 for ANY unexpected exception. Logs the type +
    correlation id only (no str(exc)/traceback: the chain may embed untrusted
    upstream strings - M1-T002 G5 F5 payload-only logging policy)."""
    return _json(
        500,
        {
            "state": "internal_error",
            "message": "unexpected internal error; see server logs by correlation id",
            "correlation_id": correlation_id,
        },
        correlation_id,
    )


def _internal_contract_error_500(correlation_id: str) -> JSONResponse:
    """Documented typed 500 for a built payload (rebuilt profile, rebuilt
    rule_evaluation, or assembled scenario) that FAILED canonical-schema
    validation before send. A payload that does not honor its contract is an
    internal defect - an invalid 200 is impossible - never a raw 500 stack."""
    return _json(
        500,
        {
            "state": "internal_contract_error",
            "message": (
                "an internal document failed canonical-contract validation and "
                "was not sent; see server logs by correlation id"
            ),
            "correlation_id": correlation_id,
        },
        correlation_id,
    )


@router.get("/properties/{bbl}/scenario", include_in_schema=False)
def get_scenario(
    bbl: str,
    fetcher: PlutoFetcher = Depends(get_pluto_fetcher),  # noqa: B008
    substrate_provider: SpatialSubstrateProvider = Depends(  # noqa: B008
        get_spatial_substrate_provider
    ),
) -> JSONResponse:
    """Rebuild the profile + rule_evaluation server-side, run the deterministic
    scenario builder, and return a scenario @ 1.0.0 document. Feature-flag gated
    OFF by default."""
    # Guard 1 (fail-safe production disable): absent/unknown flag -> 404 with no
    # hint the feature exists. Checked FIRST, before a correlation id is minted or
    # any input is touched.
    if not internal_scenario_enabled():
        return _not_found()

    correlation_id = uuid.uuid4().hex

    # 1. Validate the BBL BEFORE any connector call (typed 422; zero network I/O).
    try:
        normalized = normalize_bbl(bbl)
    except BBLValidationError as exc:
        payload = exc.to_payload()  # raw_value is repr()-sanitized there
        logger.info(
            "scenario_v1 validation_error code=%s correlation_id=%s",
            payload["code"], correlation_id,
        )
        return _json(
            422,
            {
                "state": "validation_error",
                "message": payload["message"],
                "correlation_id": correlation_id,
                "detail": {"code": payload["code"], "raw_value": payload["raw_value"]},
            },
            correlation_id,
        )

    # 2. Fetch through the injected connector; map typed failures to the SAME
    #    documented HTTP semantics as the property / rule-evaluation routes.
    try:
        result = fetcher(normalized.canonical, correlation_id)
    except PlutoConnectorError as exc:
        payload = exc.to_payload()
        logger.warning(
            "scenario_v1 connector_error state=%s correlation_id=%s",
            payload["error_type"], correlation_id,
        )
        status_code = _ERROR_STATUS.get(payload["error_type"], _DEFAULT_ERROR_STATUS)
        return _json(
            status_code,
            {
                "state": payload["error_type"],
                "message": payload["message"],
                "correlation_id": payload["correlation_id"],
                "source_id": payload["source_id"],
                "dataset_id": payload["dataset_id"],
                "detail": payload["detail"],
            },
            correlation_id,
        )
    except Exception:
        logger.error(
            "scenario_v1 unexpected_error stage=fetch correlation_id=%s", correlation_id
        )
        return _internal_error_500(correlation_id)

    # 3. Everything after the fetch runs inside one generic-500 guard so ANY
    #    unexpected exception honors the documented contract, never Starlette's
    #    plain-text 500 with full-traceback logging.
    try:
        # no_match is a RESULT, not an error: the property does not exist in the
        # official dataset, so there is nothing to build a scenario over. Same
        # 404 + machine state shape as the property / rule-evaluation routes.
        if result.status == "no_match":
            return _json(
                404,
                {
                    "state": "no_match",
                    "bbl": result.bbl,
                    "message": result.no_match_explanation,
                    "correlation_id": result.correlation_id,
                    "source_id": SOURCE_ID,
                    "dataset_id": DATASET_ID,
                    "request_url": result.request_url,
                    "retrieved_at": result.retrieved_at,
                },
                correlation_id,
            )

        # Rebuild the profile from the TRUSTED server-side path. The spatial
        # substrate comes from the injected server-side provider, never the
        # request. A None substrate is exactly the PLUTO-only build.
        substrate = substrate_provider(normalized.canonical, correlation_id)
        profile = build_property_profile(result, spatial_intersection=substrate)

        # Validate the rebuilt profile against its canonical schema before use.
        try:
            validate_profile(profile)
        except (UnsupportedContractVersionError, ContractValidationError):
            logger.error(
                "scenario_v1 profile_contract_error correlation_id=%s", correlation_id
            )
            return _internal_contract_error_500(correlation_id)

        # Rebuild the rule_evaluation document over the SAME deterministic path
        # the accepted rule-evaluation route uses (evaluate -> serialize ->
        # strict validate). A needs-review / unsupported / fail-safe evaluation
        # is a NORMAL rule_evaluation document here, never an error; the scenario
        # builder consumes it read-only.
        evaluation = evaluate_property(profile)
        rule_evaluation = serialize_rule_evaluation(
            evaluation,
            profile_contract_version=profile["profile_version"]["contract_version"],
        )
        try:
            validate_rule_evaluation_document(rule_evaluation)
        except RuleEvaluationContractError as exc:
            logger.error(
                "scenario_v1 rule_evaluation_contract_error location=%s correlation_id=%s",
                exc.location, correlation_id,
            )
            return _internal_contract_error_500(correlation_id)

        # Run the already-built deterministic scenario builder. It surfaces the
        # canonical cap VERBATIM from the trace (never recomputed) and fails
        # closed to a typed no_scenario outcome on any conflict / uncertainty /
        # malformed or absent controlling input - a NORMAL 200 document.
        scenario = build_scenario(profile, rule_evaluation)

        # Strict response validation before send: an invalid 200 is impossible;
        # the validator also fails closed on any 'verified' coverage_status.
        try:
            validate_scenario_document(scenario)
        except ScenarioContractError as exc:
            logger.error(
                "scenario_v1 scenario_contract_error location=%s correlation_id=%s",
                exc.location, correlation_id,
            )
            return _internal_contract_error_500(correlation_id)

        return _json(200, scenario, correlation_id)
    except Exception:
        logger.error(
            "scenario_v1 unexpected_error stage=build correlation_id=%s", correlation_id
        )
        return _internal_error_500(correlation_id)

File: apps/web/src/lib/scenario.ts
/**
 * HARDENED typed client + presentation classifier + feature-flag helpers for
 * the INTERNAL draft scenario surface (task M5-T002).
 *
 * Contract: services/api/app/api/v1/scenario.py (read-only dependency), the
 * flag-gated internal GET /api/v1/properties/{bbl}/scenario.
 *
 * This module transports, verifies shape, and CLASSIFIES a server-computed
 * scenario into a presentation template; it never computes a legal value, never
 * recomputes or relabels the surfaced draft zoning-floor-area cap, never decides
 * coverage, and never fills a missing envelope constraint
 * (docs/PRODUCT_FLOW_AND_AI_BOUNDARIES.md — legal logic is the deterministic
 * backend's and the scenario builder's; the frontend only displays it). Every
 * discriminator the classifier reads (scenario_kind, coverage_status,
 * professional_review_required) is produced by the backend.
 *
 * Guarantees mirrored from src/lib/rule-evaluation.ts, guarantee-for-guarantee:
 *   1. EXACT (HTTP status, body state) pair enforcement. A response whose pair
 *      is outside the documented scenario matrix renders as a distinct
 *      `unexpected_response`; a body is never routed by its `state` alone.
 *   2. Every 200 body is runtime-validated against the GENERATED canonical
 *      types (src/lib/scenario-contract.ts) BEFORE any rendering; failure is a
 *      distinct `validation_failure` carrying only a bounded problem list.
 *   3. All reflected server text is length-capped and control-stripped; the
 *      correlation id is token-allowlisted.
 *   4. Requests are cancellable (AbortController) and time-bounded; a superseded
 *      request resolves to `aborted`, a timeout to the recoverable
 *      `client_timeout`.
 *
 * The DISABLED-server case is first-class: when the endpoint is flag-gated off
 * (or unmounted) it returns a generic `404 {"detail":"Not Found"}` with no
 * `state` and no correlation id. That documented (404, null) pair maps to the
 * benign `feature_unavailable` outcome, which the UI shows as an honest "not
 * available in this environment" note that NEVER blocks the property profile.
 */

import { apiBaseUrl } from "./api";
import { boundedText, boundedToken } from "./bounded";
import { validateScenarioDocument, type Scenario } from "./scenario-contract";

/** Default request budget; kept below the Playwright test timeout so the
 * timeout journey is provable in CI without configuration. */
export const DEFAULT_TIMEOUT_MS = 12_000;

// ---------------------------------------------------------------------------
// Frontend feature flag (defense in depth; the server is independently gated).
//
// The surface is OFF by default and is gated by TWO independent conditions,
// both of which must hold:
//
//  1. ENVIRONMENT: the runtime, server-read variable INTERNAL_SCENARIO_UI holds
//     an explicit true token. The name is intentionally NOT prefixed
//     NEXT_PUBLIC_, so Next never inlines it into the browser bundle at build
//     time and never leaks the flag or the endpoint to the client: the Server
//     Component reads it once per request and passes a plain boolean into the
//     client tree. Absent / empty / unknown -> disabled (fail safe), so a
//     production deploy that never sets it keeps the surface unreachable. (This
//     mirrors rule-evaluation's INTERNAL_RULE_EVAL_UI guarantee-for-guarantee;
//     see the producer report for the deliberate non-public choice.)
//
//  2. PER-REQUEST OPT-IN: the request explicitly asks for the surface via
//     `?scenario=on`. Absent (or `off`) -> disabled. This second factor keeps
//     the experimental surface silent unless deliberately requested even where
//     the environment allows it, and lets the shared single-server e2e harness
//     enable the surface for the scenario journeys WITHOUT rendering it (or
//     issuing its fetch) on any other journey. In production, where the
//     environment gate is closed, the opt-in has no effect at all.
//
// When the resulting boolean is false the surface is never rendered and the
// scenario fetch is never issued.
// ---------------------------------------------------------------------------

const TRUE_TOKENS: ReadonlySet<string> = new Set(["1", "true", "yes", "on"]);
export const INTERNAL_SCENARIO_UI_ENV_VAR = "INTERNAL_SCENARIO_UI";

/** The env-level flag: an explicit true token enables it; absent / empty /
 * unknown -> disabled (fail safe). Read server-side only. */
export function scenarioFlagEnabled(
  rawValue: string | undefined = process.env[INTERNAL_SCENARIO_UI_ENV_VAR],
): boolean {
  return typeof rawValue === "string" && TRUE_TOKENS.has(rawValue.trim().toLowerCase());
}

/** Whether to render the scenario surface for THIS request: the env flag must
 * be on AND the request must explicitly opt in with `?scenario=on`. Default (no
 * env, no params, or `?scenario=off`) is OFF. */
export function scenarioSurfaceEnabled(params?: {
  scenario?: string | string[] | undefined;
}): boolean {
  if (!scenarioFlagEnabled()) return false;
  const raw = params?.scenario;
  const value = Array.isArray(raw) ? raw[0] : raw;
  return typeof value === "string" && TRUE_TOKENS.has(value.trim().toLowerCase());
}

// ---------------------------------------------------------------------------
// Outcome union (each documented envelope from scenario.py + the browser-level
// failure modes). Mirrors rule-evaluation's outcome union.
// ---------------------------------------------------------------------------

export interface ScenarioOutcomeDoc {
  kind: "scenario";
  document: Scenario;
  correlationId: string | null;
}

/** Endpoint flag-gated off (or unmounted): generic 404 {"detail":"Not Found"},
 * no `state`, no correlation id. Benign — never blocks the profile. */
export interface ScenarioFeatureUnavailableOutcome {
  kind: "feature_unavailable";
}

export interface ScenarioNoMatchOutcome {
  kind: "no_match";
  bbl: string | null;
  message: string;
  correlationId: string | null;
}

export interface ScenarioValidationErrorOutcome {
  kind: "validation_error";
  code: string;
  message: string;
  correlationId: string | null;
}

export const SCENARIO_UPSTREAM_FAILURE_STATES = [
  "rate_limited",
  "source_unavailable",
  "timeout",
  "schema_drift",
] as const;
export type ScenarioUpstreamFailureState = (typeof SCENARIO_UPSTREAM_FAILURE_STATES)[number];

export interface ScenarioUpstreamFailureOutcome {
  kind: "upstream_failure";
  state: ScenarioUpstreamFailureState;
  httpStatus: number;
  message: string;
  correlationId: string | null;
}

export interface ScenarioInternalErrorOutcome {
  kind: "internal_error";
  message: string;
  correlationId: string | null;
}

/** Documented 500 where the SERVER refused to ship an invalid document
 * (state=internal_contract_error). */
export interface ScenarioServerContractErrorOutcome {
  kind: "server_contract_error";
  message: string;
  correlationId: string | null;
}

/** A 200 whose body failed CLIENT-side canonical validation. */
export interface ScenarioValidationFailureOutcome {
  kind: "validation_failure";
  problems: string[];
  correlationId: string | null;
}

export interface ScenarioNetworkErrorOutcome {
  kind: "network_error";
  message: string;
}

export interface ScenarioClientTimeoutOutcome {
  kind: "client_timeout";
  timeoutMs: number;
}

export interface ScenarioAbortedOutcome {
  kind: "aborted";
}

export interface ScenarioUnexpectedResponseOutcome {
  kind: "unexpected_response";
  httpStatus: number;
  receivedState: string | null;
  correlationId: string | null;
}

export type ScenarioOutcome =
  | ScenarioOutcomeDoc
  | ScenarioFeatureUnavailableOutcome
  | ScenarioNoMatchOutcome
  | ScenarioValidationErrorOutcome
  | ScenarioUpstreamFailureOutcome
  | ScenarioInternalErrorOutcome
  | ScenarioServerContractErrorOutcome
  | ScenarioValidationFailureOutcome
  | ScenarioNetworkErrorOutcome
  | ScenarioClientTimeoutOutcome
  | ScenarioAbortedOutcome
  | ScenarioUnexpectedResponseOutcome;

/** Outcomes on which a Retry is meaningful (recoverable server/network faults;
 * the profile stays usable regardless). */
export function scenarioOutcomeIsRecoverable(outcome: ScenarioOutcome): boolean {
  return (
    outcome.kind === "upstream_failure" ||
    outcome.kind === "internal_error" ||
    outcome.kind === "server_contract_error" ||
    outcome.kind === "validation_failure" ||
    outcome.kind === "network_error" ||
    outcome.kind === "client_timeout" ||
    outcome.kind === "unexpected_response"
  );
}

// ---------------------------------------------------------------------------
// Exact (HTTP status, state) pair matrix — mirrors scenario.py verbatim (which
// mirrors the rule-evaluation route's error mapping). (404, no_match) is a
// RESULT, and the extra (404, null) pair is the flag-gated / unmounted generic
// Not Found.
// ---------------------------------------------------------------------------

type DocumentedScenarioPair = readonly [number, string | null];

const DOCUMENTED_SCENARIO_PAIRS: readonly DocumentedScenarioPair[] = [
  [200, null], // scenario document (validated client-side before render)
  [404, null], // generic Not Found: feature flag off / route unmounted
  [422, "validation_error"],
  [404, "no_match"],
  [502, "schema_drift"],
  [503, "rate_limited"],
  [503, "source_unavailable"],
  [504, "timeout"],
  [500, "internal_error"],
  [500, "internal_contract_error"],
] as const;

const SCENARIO_PAIR_KEYS: ReadonlySet<string> = new Set(
  DOCUMENTED_SCENARIO_PAIRS.map(([status, state]) => `${status}:${state ?? ""}`),
);

export function isDocumentedScenarioPair(status: number, state: string | null): boolean {
  return SCENARIO_PAIR_KEYS.has(`${status}:${state ?? ""}`);
}

export interface ScenarioLookupOptions {
  fetchImpl?: typeof fetch;
  signal?: AbortSignal;
  timeoutMs?: number;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

/**
 * Fetch and classify the internal draft scenario result for a BBL. The caller
 * must already have decided the surface is enabled (this function is never
 * invoked when the flag is off — the surface is not rendered at all). Only the
 * bbl path parameter is ever sent; no request body, no query-supplied facts.
 */
export async function fetchScenario(
  bbl: string,
  options: ScenarioLookupOptions = {},
): Promise<ScenarioOutcome> {
  const fetchImpl = options.fetchImpl ?? fetch;
  const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const url = `${apiBaseUrl()}/api/v1/properties/${encodeURIComponent(bbl)}/scenario`;

  const controller = new AbortController();
  let timedOut = false;
  const externalSignal = options.signal;
  if (externalSignal?.aborted) {
    return { kind: "aborted" };
  }
  const onExternalAbort = () => controller.abort();
  externalSignal?.addEventListener("abort", onExternalAbort);
  const timer = setTimeout(() => {
    timedOut = true;
    controller.abort();
  }, timeoutMs);

  try {
    let response: Response;
    try {
      response = await fetchImpl(url, {
        method: "GET",
        headers: { Accept: "application/json" },
        cache: "no-store",
        signal: controller.signal,
      });
    } catch {
      if (timedOut) return { kind: "client_timeout", timeoutMs };
      if (controller.signal.aborted || externalSignal?.aborted) return { kind: "aborted" };
      return {
        kind: "network_error",
        message:
          "The draft scenario service could not be reached. Nothing was assembled. " +
          "The property profile above is unaffected, and this is safe to retry.",
      };
    }

    const correlationId = boundedToken(response.headers.get("X-Correlation-ID"));

    let body: unknown = null;
    try {
      body = await response.json();
    } catch {
      if (timedOut) return { kind: "client_timeout", timeoutMs };
      if (controller.signal.aborted || externalSignal?.aborted) return { kind: "aborted" };
      return {
        kind: "unexpected_response",
        httpStatus: response.status,
        receivedState: null,
        correlationId,
      };
    }

    const record = asRecord(body);
    // RAW state for contract checks — sanitizing before comparison could launder
    // a malformed state into a documented one. Bounded form is display-only.
    const state = record && typeof record.state === "string" ? record.state : null;

    if (!isDocumentedScenarioPair(response.status, state)) {
      return {
        kind: "unexpected_response",
        httpStatus: response.status,
        receivedState: state === null ? null : boundedToken(state, 48),
        correlationId,
      };
    }

    if (response.status === 200) {
      const validation = validateScenarioDocument(body);
      if (!validation.ok) {
        return {
          kind: "validation_failure",
          problems: validation.problems.map((problem) =>
            boundedText(problem, "problem detail unavailable"),
          ),
          correlationId,
        };
      }
      return { kind: "scenario", document: validation.document, correlationId };
    }

    // (404, null): generic Not Found — the feature is disabled or unmounted.
    if (response.status === 404 && state === null) {
      return { kind: "feature_unavailable" };
    }

    if (state === "no_match") {
      return {
        kind: "no_match",
        bbl: typeof record?.bbl === "string" ? boundedToken(record.bbl, 32) : null,
        message: boundedText(
          record?.message,
          "No record was found for this BBL in the current official dataset.",
        ),
        correlationId,
      };
    }

    if (state === "validation_error") {
      const detail = asRecord(record?.detail);
      return {
        kind: "validation_error",
        code: boundedToken(detail?.code, 48) ?? "unknown",
        message: boundedText(record?.message, "The BBL was rejected by the API."),
        correlationId,
      };
    }

    if ((SCENARIO_UPSTREAM_FAILURE_STATES as readonly string[]).includes(state ?? "")) {
      return {
        kind: "upstream_failure",
        state: state as ScenarioUpstreamFailureState,
        httpStatus: response.status,
        message: boundedText(record?.message, "The official data source failed."),
        correlationId,
      };
    }

    if (state === "internal_contract_error") {
      return {
        kind: "server_contract_error",
        message: boundedText(
          record?.message,
          "The server refused to deliver a draft scenario that failed its contract checks.",
        ),
        correlationId,
      };
    }

    // Only (500, internal_error) remains in the documented matrix.
    return {
      kind: "internal_error",
      message: boundedText(record?.message, "Unexpected internal error."),
      correlationId,
    };
  } finally {
    clearTimeout(timer);
    externalSignal?.removeEventListener("abort", onExternalAbort);
  }
}

// ---------------------------------------------------------------------------
// Presentation classifier — maps a server-computed scenario document onto
// exactly one of the honest UI states. It reads ONLY backend discriminators
// (scenario_kind, coverage_status, professional_review_required) and performs
// NO legal reasoning; the priority order picks the most specific honest framing.
// ---------------------------------------------------------------------------

export type ScenarioPresentation =
  | "preliminary_cap" // preliminary scenario: the draft cap is surfaced verbatim
  | "unsupported" // no implemented rule family for this district
  | "conflict" // competing rules / conflicting data; no value
  | "professional_review" // spatial / professional-review uncertainty; no value
  | "missing"; // a required controlling input is absent; no value

export function classifyScenario(document: Scenario): ScenarioPresentation {
  if (document.scenario_kind === "preliminary") return "preliminary_cap";
  const coverage = document.coverage_status;
  if (
    document.scenario_kind === "unsupported" ||
    coverage === "unsupported" ||
    coverage === "not_applicable"
  ) {
    return "unsupported";
  }
  if (coverage === "data_conflict") return "conflict";
  if (document.professional_review_required || coverage === "professional_review_required") {
    return "professional_review";
  }
  return "missing";
}

// ---------------------------------------------------------------------------
// Assistive-technology announcement copy for scenario outcome arrivals. Derived
// deterministically from the already-classified outcome; no legal semantics, no
// "verified"/"best"/"guaranteed" wording, no invented values. `aborted`
// announces nothing (a superseded request).
// ---------------------------------------------------------------------------

const SCENARIO_UPSTREAM_ANNOUNCEMENTS: Record<ScenarioUpstreamFailureState, string> = {
  rate_limited: "Draft scenario unavailable: the official data source is throttling requests.",
  source_unavailable: "Draft scenario unavailable: the official data source is unavailable.",
  timeout: "Draft scenario unavailable: the official data source timed out.",
  schema_drift: "Draft scenario unavailable: the official dataset changed shape.",
};

const SCENARIO_PRESENTATION_ANNOUNCEMENTS: Record<ScenarioPresentation, string> = {
  preliminary_cap:
    "Draft scenario loaded: a preliminary draft zoning-floor-area cap that requires professional review; not a buildable envelope.",
  unsupported:
    "Draft scenario loaded: no draft rule applies to this property, so no scenario was produced.",
  conflict:
    "Draft scenario loaded: conflicting draft rules or data; professional review required and no value produced.",
  professional_review:
    "Draft scenario loaded: professional review required; spatial uncertainty blocks a draft scenario and no value was produced.",
  missing:
    "Draft scenario loaded: a required controlling input is missing, so no draft scenario value was produced.",
};

export function announcementForScenario(outcome: ScenarioOutcome): string {
  switch (outcome.kind) {
    case "scenario":
      return SCENARIO_PRESENTATION_ANNOUNCEMENTS[classifyScenario(outcome.document)];
    case "feature_unavailable":
      return "Draft scenario is not available in this environment.";
    case "no_match":
      return "Draft scenario: no property record found in the official dataset.";
    case "validation_error":
      return "Draft scenario rejected: the API rejected this BBL.";
    case "upstream_failure":
      return SCENARIO_UPSTREAM_ANNOUNCEMENTS[outcome.state];
    case "internal_error":
      return "Draft scenario failed: something went wrong on our side.";
    case "server_contract_error":
      return "Draft scenario failed: the server refused to deliver an invalid draft document.";
    case "validation_failure":
      return "Draft scenario failed: the response did not match the published data contract.";
    case "network_error":
      return "Draft scenario failed: the service could not be reached. The property profile is unaffected.";
    case "client_timeout":
      return "Draft scenario failed: the request took too long and was cancelled.";
    case "unexpected_response":
      return "Draft scenario failed: unexpected response from the platform API.";
    case "aborted":
      return "";
  }
}
